contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,564 @@
+Spec:
+  volumeType: block
+  capacity:
+    storage: "10Gi"
+  accessModes:
+    - "ReadWriteOnce"
+  gcePersistentDisk:
+    fsType: "block"
+    pdName: "gce-disk-1"
+```
+
+***If admin specifies volumeType: block + fstype: ext4 then they would have the default behavior of files on block ***
+***fsType values will be provisioner dependent. Block is suggested for development simplicity. Since the PVC object is passed
+   to the provisioner, it will be responsible for validating and handling whether or not it supports the volumeType being passed ***
+
+# Container Runtime considerations
+It is important the values that are passed to the container runtimes are valid and support the current implementation of these various runtimes. Listed below are a table of various runtime and the mapping of their values to what is passed from the mount. 
+
+| runtime engine    | runtime options  | accessMode       |  
+| --------------    |:----------------:| ----------------:|
+| docker/runc/rkt   |  mknod / RWM     | RWO              |
+| docker/runc/rkt   |       R          | ROX              |
+
+The accessModes would be passed as part of the options array and would need validate against the specific runtime engine. 
+Since rkt doesn't use the CRI, the config values would need to be passed in the legacy method.
+
+The runtime option would be placed in the DeviceInfo as such:
+devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: path, Permissions: "mrw"}) for RWO
+devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: path, Permissions: "mr"}) for ROX
+
+Today, this is defaulted always to mrw, thus it would need to be updated with what is passed in.
+
+# Implementation Plan, Features & Milesones
+
+Phase 1: v1.8
+Feature: Pre-provisioned PVs to precreated devices 
+
+               Milestone 1: API changes
+
+               Milestone 2: Restricted Access 
+               
+               Milestone 3: Scrubbing device after PV is deleted
+               
+               Milestone 4: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 5: Expose volumeType to users via kubectl
+               
+               Milestone 6: Adds enable/disable configuration to securityContext in PSP (Pod Security Policy) similar to hostPath
+               
+               Milestone 7: Validate container runtime options with user specifcations as indicated in UC3
+	       
+	       Milestone 8: Container Runtime changes
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
```

Comment:
(starting a new thread in arbitrary place)

### External provisioners

#### Intro

We have a "protocol" to allow dynamic provisioning by external software (pods, daemons running anywhere, ...) Full details are here: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/volume-provisioning.md

tl;dr: the protocol is very simple. External provisioner watches PVCs in API server and when it sees a PVC with a special annotation (`volume.beta.kubernetes.io/storage-provisioner == <the provisioner name>`), it provisions a PV for it and stores it into API server (or sends an error event to the PVC so user knows that something went wrong). PV controller then binds them together and user is happy. **There is no gPRC or any other direct API call between PV controller and the external provisioner. Everything happens through annotations on PVCs.**

#### Problem

Existing external provisioners don't know about new fields that we add to PVs, PVCs or StorageClasses, e.g. `pvc.volumeType`. The provisioner gets a PVC, inspects the fields it knows about, ignores `volumeType` and provisions a PV with empty `volumeType`. That's bad when user asked for `volumeType=block`, because (implicit) `file` is provisioned.

This raw block volume PR should extend volume-provisioning.md to include generic extension mechanism which we will use to add new features to PVs/PVCs/StorageClasses where legacy provisioners (those that do not know about `pvc.volumeType`) can still provision file PVs and new provisioners can provision both fine and block PVs.

#### Proposal

PV controller:
* `pvc.volumeType == file`: both legacy and new provisioners can provision a PV. PV controller puts these annotations to the PVC:
  * `volume.beta.kubernetes.io/storage-provisioner = <provisioner name>`
  * `volume.kubernetes.io/storage-provisioner-with-features = <provisioner name>`
  * `volume.kubernetes.io/storage-provisioning-features = VolumeType`

* `pvc.volumeType == block`: only new provisioners can provision a PV. PV controller puts these annotations to the PVC:
  * `volume.kubernetes.io/storage-provisioner-with-features = <provisioner name>`
  * `volume.kubernetes.io/storage-provisioning-features = VolumeType`

Legacy provisioner (no changes are required here, this behavior below is already implemented):

* `pvc.volumeType == file`: The provisioner sees `volume.beta.kubernetes.io/storage-provisioner = <provisioner name>` and provisions a PV as usual. It ignores any other annotations.
* `pvc.volumeType == block`: The provisoner does not see any `volume.beta.kubernetes.io/storage-provisioner` annotation and ignores the PVC. The PVC is `Pending` forever.

New provisioner:

* The provisioner sees `volume.kubernetes.io/storage-provisioner-with-features = <provisioner name>` and inspects `volume.kubernetes.io/storage-provisioning-features` annotation that contains comma-separated list of "features". VolumeType is the only known feature for now, but let's assume we add feature called Encryption with a new 1st class field `pvc.encryptionKey` in the future.
  * If an unknown feature is there, the provisioner puts an error event to the PVC so user knows, say "Unknown feature specified: Encryption". 
  * If the provisioner supports all features in the annotation, it provisions the PV as usual. E.g. for feature VolumeType the provisioner knows that `pvc.volumeType` should be evaluated, for feature Encryption it knows how to deal with `pvc.encryptionKey` and so on.

With every new feature (new behavior expected, new fields in PV/PVC/StorageClass), we create a new Feature and document it somewhere.


#### Alternative solution
Using version instead of list of features: the protocol would be almost the same as above, just with `storage-provisioning-version` annotation instead of comma-separated list in `storage-provisioning-features` annotation. Somewhere in Kubernetes docs we would describe what each version means, e.g. :
  * V1 adds support for `pvc.volumeType` (i.e. anyone who accepts PVC with version V1 in the annotation supports `pvc.volumeType`)
  * V2 adds pvc.encryption (i.e. anyone who accepts PVC with version V2 in the annotation knows how to handle `pvc.encryptionKey` **and** all older versions, i.e. `pvc.volumeType` )

I like the comma-separated list of features more - it's more obvious what the provisioner needs to support to satisfy the claim. On the other hand, the list may get very long over the time... I hope CSI saves us here.

#### Implementation

Implementation in PV controller is quite simple - just add the annotations to PVCs as describe above.
Legacy provisioners don't need any change. New provisioners need some updates, bu you can see it's not that complicated. We'll updated provisioners in [external-storage](https://github.com/kubernetes-incubator/external-storage/) accordingly.

The biggest pain point is update of volume-provisioning.md.